File: MastersProject/resteaze_dash.py
import sys
import os
import numpy as np
import datetime
from numpy import sqrt, square
from utilities import Param, Output, nightData, WASO, sleepText, rms, find
from syncRE import syncRE
from getLMiPod import getLMiPod
from candidate_lms import candidate_lms
from scoreSleep import scoreSleep
from calculateArousal import calculateArousal
from periodic_lms import periodic_lms
from calculateWASO_RE import calculateWASO_RE


def resteaze_dash(left,right,subjectid):

    np.seterr(divide='ignore', invalid='ignore')

    params = init_params()
    output = init_output(subjectid)

    if os.path.exists(left):
        leftPath = os.path.splitext(left)[0]
        #leftFileNames = os.path.splitext(left)[1]
        ext = os.path.splitext(left)[1]

        print("leftpath: "+leftPath)
        #print("leftFileNames: "+leftFileNames)
    else:
        print("lefterror")
    if os.path.exists(right):
        
        rightPath = os.path.splitext(right)[0]
        #rightFileNames = os.path.splitext(right)[1]
        right_ext = os.path.splitext(right)[1]
        print("rightpath: " + rightPath)
        #print("rightFileNames: " + rightFileNames)

    else:
        print("righterror")

    
    #add code to read and process multiple csv's

    """ read data from csv files """
    bandData_left = np.genfromtxt(left,delimiter=',',skip_header=1)
    bandData_right = np.genfromtxt(right,delimiter=',',skip_header=1)

    """synching signals from two legs"""
    leftLeg,rightLeg = syncRE(bandData_left,bandData_right)

    output.up2Down1 = np.ones((leftLeg.shape[0],1)) 
    
    #################################################

    """calculating root-mean-square of the acclerometer movements for both legs"""
    output.lRMS = rms(leftLeg[:,[1,2,3]])
    output.rRMS = rms(rightLeg[:,[1,2,3]])

    #################################################
    
    """ compute LM(leg movements) """
    rLM = getLMiPod(params,output.rRMS,output.up2Down1)
    lLM = getLMiPod(params,output.lRMS,output.up2Down1)
    
    #################################################

    """ Start Patrick's standard scoring stuff """
    bCLM = candidate_lms(rLM,lLM,params)
    Arousal = calculateArousal(bCLM,leftLeg,rightLeg)
    output.Arousal = Arousal[Arousal[:,2]==1,:]
    bCLM[:,10] = Arousal[:,2]
    PLM, bCLM = periodic_lms(bCLM,params)
    #################################################
    PLM = np.asarray(PLM)
    
    """  score sleep/wake """
    output.wake = scoreSleep(params.fs,output.lRMS,PLM,bCLM)
    WASO = calculateWASO_RE(output.wake,params.minSleepTime,params.fs) # NEED TO UPDATE OUTPUT MATRICES AFTER HERE
    
    rLM[:,5] = output.wake[np.array(rLM[:,0],dtype=int)] + 1 # still using up2down1 format
    lLM[:,5] = output.wake[np.array(lLM[:,0],dtype=int)] + 1 # still using up2down1 format
    bCLM[:,5] = output.wake[np.array(bCLM[:,0],dtype=int)] + 1 # still using up2down1 format
    
    if len(PLM) != 0:
        PLM[:,5] = output.wake[np.array(PLM[:,0],dtype=int)] + 1  # still using up2down1 format
    
    else:
        PLM = []
    #################################################

    """ Matrices to output """
    output.rLM = rLM
    output.lLM = lLM
    
    output.bCLM = bCLM
    output.PLM = PLM
    output.PLMS = []
    if len(PLM) != 0:
        xx = find(PLM[:,5],lambda x: x==1)
        output.PLMS = PLM[xx,:]
    
    #################################################
    """ Quantitative measures to output """

    output.TRT = (output.up2Down1.shape[0]/params.fs)/60
    output.TST = output.TRT - WASO.dur
    output.sleepEff = output.TST / output.TRT
    output.WASOnum = WASO.num
    output.WASOdur = WASO.dur
    output.WASOavgdur = WASO.avgdur


    if len(output.PLMS) != 0: #change this to !=0
        output.PI = sum( np.array(output.PLMS[:,8]==0,dtype=int) ) / (output.bCLM.shape[0]-1)
        output.PLMSArI = sum(output.PLMS[:,10])/(output.TST/60) # num plms arousals per hr of sleep

    if len(PLM) != 0:
        plm_5 = PLM[:,5] == 1 
        plm_8 = PLM[:,8] == 0
        plm_combi = PLM[plm_5|plm_8,3]

        output.avglogPLMSIMI = np.mean(np.log(plm_combi))
        output.stdlogPLMSIMI = np.std(np.log(plm_combi))

        output.avgPLMSDuration = np.mean(PLM[plm_5,2])
        output.stdPLMSDuration = np.std(PLM[plm_5,2])
    
        output.PLMhr = PLM.shape[0]/(output.TRT/60)
        output.PLMShr = sum(PLM[:,5]==1)/(output.TST/60)
        output.PLMWhr = sum(PLM[:,5]==2)/((output.TRT-output.TST)/60)

        output.PLMnum = len(PLM)
        output.PLMSnum = sum(PLM[:,5]==1)
        output.PLMWnum = sum(PLM[:,6]==2)

    if len(bCLM) != 0:
        output.avglogCLMSIMI = np.mean(np.log(bCLM[bCLM[:,5]==1,3]))
        output.stdlogCLMSIMI = np.std(np.log(bCLM[bCLM[:,5]==1,3]))

        output.avgCLMSDuration = np.mean(bCLM[bCLM[:,5]==1,2])
        output.stdCLMSDuration = np.std(bCLM[bCLM[:,5]==1,2])

        output.CLMhr = bCLM.shape[0]/(output.TRT/60)
        output.CLMShr = sum(bCLM[:,5]==1)/(output.TST/60)
        output.CLMWhr = sum(bCLM[:,5]==2)/((output.TRT-output.TST)/60)

        output.CLMnum = bCLM.shape[0]
        output.CLMSnum = sum(bCLM[:,5]==1)
        output.CLMWnum = sum(bCLM[:,5]==2)

        output.GLM = bCLM[bCLM[:,4]==0,:]

    #print("avglogCLMSIMI ",output.avglogCLMSIMI)
    #print("stdlogCLMSIMI ",output.stdlogCLMSIMI)
    #print("avgCLMSDuration ",output.avgCLMSDuration)
    #print("stdCLMSDuration ",output.stdCLMSDuration)
    
    #print("avglogPLMSIMI ",output.avglogPLMSIMI)
    #print("stdlogPLMSIMI ",output.stdlogPLMSIMI)
    #print("avglogPLMSIMI ",output.avgPLMSDuration)
    #print("stdlogPLMSIMI ",output.stdPLMSDuration)
    
    #print("CLMhr ",output.CLMhr)
    #print("CLMShr ",output.CLMShr)
    #print("CLMWhr ",output.CLMWhr)
    
    #print("PLMhr ",output.PLMhr)
    #print("PLMShr ",output.PLMShr)
    #print("PLMWhr ",output.PLMWhr)
    
    #print("CLMnum ",output.CLMnum)
    #print("CLMSnum ",output.CLMSnum)
    #print("CLMWnum ",output.CLMWnum)        
    
    #print("PLMSArI ",output.PLMSArI)
    #print("GLM ",output.GLM.shape)

    #some more stuff needed for report 
    output.intervalSize = 1
    output.fs = params.fs
    output.pos = 2 * np.ones(leftLeg.shape[0]) # ALL BACKSIDE RIGHT NOW SINCE NO POS VECTOR YET
    output.ArI = output.Arousal.shape[0]/(output.TST/60)
    output.PLMSI = output.PLMShr
    leftStart = datetime.datetime.fromtimestamp(leftLeg[0,3] / 1000)
    leftStart = [leftStart.year, leftStart.month, leftStart.day, leftStart.hour, leftStart.minute, leftStart.second]
    output.sleepStart = np.mod(leftStart[3]*60 + leftStart[4] - 12*60 , 24*60)
    output.sleepEnd = output.sleepStart + output.TRT
    output.date = str(leftStart[1])+ '/' +str(leftStart[2])+ '/'+ str(np.mod(leftStart[0],100))

    output.SQ = 0
    output.SQhrs = []

    nightData = output

    fileID = open(output.fileName+'.txt','w')
    fileID.write('PatientID: '+ output.fileName+'\n')
    fileID.write('Record Start: '+ sleepText(nightData.sleepStart)+'\n')
    fileID.write('Record Stop: ' + sleepText(nightData.sleepEnd)+'\n')
    fileID.write('Sleep Efficiency:  '+ '{:.2f}'.format(nightData.sleepEff)+'\n')
    fileID.write('Total Sleep Time:  '+ '{:.2f}'.format(nightData.TST)+'\n')
    fileID.write('PLMS/hr:  '+ '{:.2f}'.format(nightData.PLMShr)+'\n')
    fileID.write('Arousals/hr:  '+ '{:.2f}'.format(nightData.ArI)+'\n')
    fileID.write('Sleep Quality:  ' + '{:.2f}'.format(nightData.SQ)+'\n')
    fileID.write('WASO: '+ '{:.2f}'.format(nightData.WASOdur)+'\n')
    fileID.write('Sleep_quality_per_hr :')
    for i in range(len(nightData.SQhrs)):
        fileID.write(fileID+': '+'{:.2f}'.format(nightData.SQhrs[0,i])+'\n')
    fileID.close()

    print('PatientID: '+ output.fileName)
    print('Record Start: '+ sleepText(nightData.sleepStart))
    print('Record Stop: ' + sleepText(nightData.sleepEnd))
    print('Sleep Efficiency:  '+ str(nightData.sleepEff))
    print('Total Sleep Time:  '+ str(nightData.TST))
    print('PLMS/hr:  '+ str(nightData.PLMShr))
    print('Arousals/hr:  '+ str(nightData.ArI))
    print('Sleep Quality:  ' + str(nightData.SQ))
    print('WASO: '+ str(nightData.WASOdur))
    """
    fprintf(fileID,'Sleep_quality_per_hr : ')
    for ii=1:size(nightData(i).SQhrs,2)
        fprintf(fileID, '%.2f ', nightData(i).SQhrs(1,ii));
    end"""
    
def init_output(subjectid):
    output = Output()
    output.fileName = subjectid
    output.PI = 0

    output.avglogCLMSIMI = None
    output.stdlogCLMSIMI = None
    output.avgCLMSDuration = None
    output.stdCLMSDuration = None

    output.avglogPLMSIMI = None
    output.stdlogPLMSIMI = None
    output.avgPLMSDuration = None 
    output.stdPLMSDuration = None
    
    output.CLMhr = 0
    output.CLMShr = 0
    output.CLMWhr = 0
    output.PLMhr = 0
    output.PLMShr = 0
    output.PLMWhr = 0

    output.CLMnum = 0
    output.CLMSnum = 0
    output.CLMWnum = 0
    output.PLMnum = 0
    output.PLMSnum = 0
    output.PLMWnum = 0

    output.PLMSArI = 0
    return output

def init_params():
    # Initialize default parameters, if we want to allow them to change, modify here.
    params = Param()

    params.iLMbp='on'
    params.morphologyCriteria='on'

    # this is for the RestEaZe data.
    #params.lowThreshold=0.005;
    #params.highThreshold=0.01;
    params.lowThreshold = 0.05
    params.highThreshold = 0.1

    params.minLowDuration = 0.5
    params.minHighDuration = 0.5
    params.minIMIDuration = 10
    params.maxIMIDuration = 90
    params.maxCLMDuration = 10
    
    #sampling rate for the RestEaZe data
    #params.fs=50;
    params.fs = 25
    params.minNumIMI = 3
    params.minSleepTime = 5
    params.maxOverlapLag = 0.5
    params.maxbCLMOverlap = 4
    params.maxbCLMDuration = 15
    params.side = 'both'
    print("params initialized...")

    return params

if __name__ == '__main__':
    resteaze_dash(sys.argv[1],sys.argv[2],sys.argv[3])

File: matlabcode/resteaze_dash.m
function output = resteaze_dash(left, right, subjectid)

% Add Paths
addpath subroutines
addpath RestEaze_11_09
addpath 'Calculate Arousal'
addpath 'export_fig'

warning('off','all')
left = convertCharsToStrings(left);
right = convertCharsToStrings(right);

% Check for File exixtence
if exist('left', 'var') && exist(left, 'file') == 2
     % File exists.
     [filepath,name,ext] = fileparts(left);
     leftFileNames = strcat(name, ext);
     leftPath = filepath;
else
     % File does not exist.
     leftFileNames = [];
     leftPath = [];
end

if exist('right', 'var') && exist(right, 'file') == 2
     % File exists.
     [filepath,name,ext] = fileparts(right);
     rightFileNames = strcat(name, ext);
     rightPath = filepath;
else
     % File does not exist.
     rightFileNames = [];
     rightPath= [];
end

% scoring options
params = init_params();

% arrays and stats to output and generate report
output=struct();

%[output.fileName,path]=uigetfile('*.csv');

% TODO: obv make a better file name.
output.fileName=subjectid;

% Should have date in beginning of filename so they are sorted from old->new
if isempty(leftFileNames)
    [leftFileNames,leftPath]=uigetfile('*.csv','LEFT FILES','multiselect','on'); 
    if ~iscell(leftFileNames)
        if size(leftPath,2)==1
            if leftPath==0
                params.side='right';
                disp('Right Leg Only')
            end
        else
            leftFileNames={leftFileNames}; %Inserts single fileName str into array
        end
    end
end


if isempty(rightFileNames)
    [rightFileNames,rightPath]=uigetfile('*.csv','RIGHT FILES','multiselect','on');
    if ~iscell(rightFileNames)
        if size(rightPath,2)==1
            if rightPath==0
                if strcmp(params.side,'both')
                    params.side='left';
                    disp('Left Leg Only')

                elseif strcmp(params.side,'right')
                    error('no data');
                end
            end
        else
            rightFileNames={rightFileNames}; %Inserts single fileName str into array
        end
    end
end


clear nightData

% it appears this will fail if there are not the same number of nights for
% each leg. TODO: make this more safe
for i=1:max([size(leftFileNames,2) size(rightFileNames,2)])
    
    if strcmp(params.side,'both')
        bandData_left=readtable(fullfile(leftPath,leftFileNames{1,i}));
        bandData_right=readtable(fullfile(leftPath,rightFileNames{1,i}));
    elseif    strcmp(params.side,'right')
        bandData_right=readtable(fullfile(rightPath,rightFileNames{1,i}));
        bandData_left=readtable(fullfile(rightPath,rightFileNames{1,i}));
    elseif strcmp(params.side,'left')
        bandData_right=readtable(fullfile(leftPath,leftFileNames{1,i}));
        bandData_left=readtable(fullfile(leftPath,leftFileNames{1,i}));
    else
        error('Which side?')
    end
    leftdata=table2array(bandData_left);
    rightdata=table2array(bandData_right);
    [leftLeg,rightLeg]=synchRE(leftdata,rightdata);
    
    output.up2Down1=ones(size(leftLeg,1),1); %%%%%%%%%%%%%%%%%% ALL DOWN FOR NOW WILL HAVE TO UPDATE LATER

    output.rRMS=rms(rightLeg(:,2:4),2);
    output.lRMS=rms(leftLeg(:,2:4),2);
    
    %factor = 8/32767; % factor for conversion of accelerometer value in terms of g. 
    % the dynamic range is (+/-) 8g and the ADC is 16 bit.
    %output.rRMS=rms(removeAccGrav(rightLeg(:,4:6))*factor,2);
    %output.lRMS=rms(removeAccGrav(leftLeg(:,4:6))*factor,2);
    
    %output.rRMS=rms(rightLeg(:,4:6)*factor,2);
    %output.lRMS=rms(leftLeg(:,4:6)*factor,2);
    
    %% Compute LM:
    % Activity above high threshold, ending when there is continuous 0.5s below low threshold
    rLM=getLMiPod(params,output.rRMS,output.up2Down1);
    lLM=getLMiPod(params,output.lRMS,output.up2Down1);
    
    %% Start Patrick''s standard scoring stuff %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    bCLM = candidate_lms(rLM,lLM,params);
    Arousal=calculateArousal(bCLM,leftLeg,rightLeg);
    output.Arousal=Arousal(Arousal(:,3)==1,:);
   
    bCLM(:,11)=Arousal(:,3);
    [PLM, bCLM] = periodic_lms(bCLM,params);
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    %% score sleep/wake
    output.wake=scoreSleep(params.fs,output.lRMS,PLM,bCLM);
    [WASO] = calculateWASO_RE(output.wake,params.minSleepTime,params.fs); %%% NEED TO UPDATE OUTPUT MATRICES AFTER HERE
    rLM(:,6)=output.wake(rLM(:,1),1)+1; %still using up2down1 format
    lLM(:,6)=output.wake(lLM(:,1),1)+1; %still using up2down1 format
    bCLM(:,6)=output.wake(bCLM(:,1),1)+1; %still using up2down1 format
    PLM(:,6)=output.wake(PLM(:,1),1)+1; %still using up2down1 format
    
    
    
    
    %% Matrices to output
    output.rLM=rLM;
    output.lLM=lLM;
    
    output.bCLM=bCLM;
    output.PLM=PLM;
    output.PLMS=PLM(PLM(:,6)==1,:);
    
    %% Quantitative measures to output
    output.TRT =(size(output.up2Down1,1)/params.fs)/60; %Total rest time in mins
    output.TST = output.TRT-WASO.dur; %Total sleep time in mins
    output.sleepEff=output.TST/output.TRT;
    output.WASOnum=WASO.num;
    output.WASOdur=WASO.dur;
    output.WASOavgdur=WASO.avgdur;
    output.PI=sum((output.PLMS(:,9)==0)/(size(output.bCLM,1)-1));
    
    output.avglogCLMSIMI=mean(log(bCLM(bCLM(:,6)==1,4)));
    output.stdlogCLMSIMI=std(log(bCLM(bCLM(:,6)==1,4)));
    
    output.avgCLMSDuration=mean(bCLM(bCLM(:,6)==1,3));
    output.stdCLMSDuration=std(bCLM(bCLM(:,6)==1,3));
    
    output.avglogPLMSIMI=mean(log(PLM(PLM(:,6)==1|PLM(:,9)==0,4)));
    output.stdlogPLMSIMI=std(log(PLM(PLM(:,6)==1|PLM(:,9)==0,4)));
    
    output.avgPLMSDuration=mean(PLM(PLM(:,6)==1,3));
    output.stdPLMSDuration=std(PLM(PLM(:,6)==1,3));
    
    output.CLMhr = size(bCLM,1)/(output.TRT/60);
    output.CLMShr = sum(bCLM(:,6)==1)/(output.TST/60);
    output.CLMWhr = sum(bCLM(:,6)==2)/((output.TRT-output.TST)/60);
    
    output.PLMhr = size(PLM,1)/(output.TRT/60);
    output.PLMShr=sum(PLM(:,6)==1)/(output.TST/60);
    output.PLMWhr=sum(PLM(:,6)==2)/((output.TRT-output.TST)/60);
    
    output.CLMnum = size(bCLM,1);
    output.CLMSnum = sum(bCLM(:,6)==1);
    output.CLMWnum = sum(bCLM(:,6)==2);
    
    output.PLMnum = size(PLM,1);
    output.PLMSnum=sum(PLM(:,6)==1);
    output.PLMWnum=sum(PLM(:,6)==2);
    
    output.PLMSArI=sum(output.PLMS(:,11))/(output.TST/60); %num plms arousals per hr of sleep
    
    output.GLM=bCLM(bCLM(:,5)==0,:);
    
    %some more stuff needed for report %%%%%%%%%%%
    output.intervalSize=1;
    output.GLM=bCLM(bCLM(:,5)==0,:);
    output.Arousal=output.Arousal;
    output.fs=params.fs;
    output.pos=2*ones(size(leftLeg,1),1); %%%%%%%%%%%%%%%%%%%%% ALL BACKSIDE RIGHT NOW SINCE NO POS VECTOR YET
    output.ArI=size(output.Arousal,1)/(output.TST/60);
    output.PLMSI=output.PLMShr;
    warning('off','all') %suppress timezonewarning
    leftStart=datevec(datetime(leftLeg(1,4)/1000,'ConvertFrom','PosixTime','TimeZone','EST'));
    warning('on','all')
    output.sleepStart=mod(leftStart(1,4)*60+leftStart(1,5)-12*60,24*60); %% took mod for sleep before noon
    output.sleepEnd=output.sleepStart+output.TRT;
    output.date=[num2str(leftStart(1,2)) '/' num2str(leftStart(1,3)) '/' num2str(mod(leftStart(1,1),100))];
    output.SQ=0;
    output.SQhrs=[];

    nightData(i)=output;
    
end


%Generate report
%not called only used for debugging.
if (false)
    for i=1:ceil(size(nightData,2)/5)
        if (i-1)*5+5<=size(nightData,2)
            [Handles,nightData((i-1)*5+1:(i-1)*5+5)]=generateClinicalReport_RE(nightData((i-1)*5+1:(i-1)*5+5));
            export_fig 'graphs.pdf' -transparent;
            figure_list{1,1} = 'graphs.pdf';

        else
            [Handles,nightData((i-1)*5+1:end)]=generateClinicalReport_RE(nightData((i-1)*5+1:end));
            export_fig 'graphs.pdf' -transparent;
            figure_list{1,1} = 'graphs.pdf';

        end

        %% Combine all figures in list into one pdf
         append_pdfs([output.fileName '.pdf'], figure_list{1,:});
    end
end 

%% Write to text file

for i=1:size(nightData,2)
    fileID = fopen([ output.fileName num2str(i) '.txt'],'w');
    fprintf(fileID,'PatientID: %s\n',output.fileName);
    fprintf('PatientID: %s\n',output.fileName);
    fprintf(fileID,['Record Start: ' sleepText(nightData(i).sleepStart) '\n']);
    fprintf(['Record Start: ' sleepText(nightData(i).sleepStart) '\n']);
    fprintf(fileID,['Record Stop: ' sleepText(nightData(i).sleepEnd) '\n']);
    fprintf(['Record Stop: ' sleepText(nightData(i).sleepEnd) '\n']);
    fprintf(fileID,'Sleep Efficiency: %.2f\n',nightData(i).sleepEff);
    fprintf('Sleep Efficiency: %.2f\n',nightData(i).sleepEff);
    fprintf(fileID,'Total Sleep Time: %.2f\n',nightData(i).TST);
    fprintf('Total Sleep Time: %.2f\n',nightData(i).TST);
    fprintf(fileID,'PLMS/hr: %.2f\n',nightData(i).PLMShr);
    fprintf('PLMS/hr: %.2f\n',nightData(i).PLMShr);
    fprintf(fileID,'Arousals/hr: %.2f\n',nightData(i).ArI);
    fprintf('Arousals/hr: %.2f\n',nightData(i).ArI);
    fprintf(fileID,'Sleep Quality: %.2f\n',nightData(i).SQ);
    fprintf('Sleep Quality: %.2f\n',nightData(i).SQ);
    fprintf(fileID,'WASO: %.2f\n',nightData(i).WASOdur);
    fprintf('WASO: %.2f\n',nightData(i).WASOdur);
    fprintf(fileID,'Sleep_quality_per_hr : ');
    fprintf('Sleep_quality_per_hr : ');
    for ii=1:size(nightData(i).SQhrs,2)
        fprintf(fileID, '%.2f ', nightData(i).SQhrs(1,ii));
        fprintf('%.2f ', nightData(i).SQhrs(1,ii));
    end
    fclose(fileID);

end

end


function params = init_params()
% Initialize default parameters, if we want to allow them to change,
% modify here.
params = struct();

params.iLMbp='on';
params.morphologyCriteria='on';

%% this is for the RestEaZe data.
%params.lowThreshold=0.005;
%params.highThreshold=0.01;
params.lowThreshold=0.05;
params.highThreshold=0.1;


params.minLowDuration=0.5;
params.minHighDuration=0.5;
params.minIMIDuration=10;
params.maxIMIDuration=90;
params.maxCLMDuration=10;

% params.fs=50;
%sampling rate for the RestEaZe data
params.fs=25;

params.minNumIMI=3;
params.minSleepTime=5;
params.maxOverlapLag=0.5;
params.maxbCLMOverlap=4;
params.maxbCLMDuration=15;
params.side='both';

end